### ISSUE TITLE
Add priority on scheduling tasks

### ISSUE THREAD

**Issue by dkhmelenko** [bobvanderlinden/probot-auto-merge#92]
What do you think about adding some priority on scheduling tasks:
1. if the PR is green, approved and satisfies all the other conditions, then it should be merged right away
2. if the PR is approved and satisfies all other conditions, but out-of-date, it needs to be updated right away (if Update Branch option enabled)
3. in all other cases the PR should be scheduled in normal order in the queue.

**Comment by dkhmelenko** [comment]
This can be achieved pretty easy with `if-else` clause. I think you don't need to have a check for multiple PRs, just one check on events if it satisfies all criteria or not.

Then this will simply solve the case when a PR is ready to be merged (all checks passed etc.), but is not merged just because it is at the end of the queue.

**Comment by dkhmelenko** [comment]
> It might be worth the effort though, as I do like immediately handling PRs that have a green merge button. But currently I'm still a bit on the fence for this one.

What kind of concerns do you have for that?

> This will not fly I think. Say there are 10 PRs open, all ready for merging and the repository has strict protected branches enabled (a very sane configuration when used combined with auto-merge). Whenever one PR gets merged, all other PRs will be out-of-date and need to be updated. Whenever all PRs will be updated at once, 9 different builds will be run. Once the first PR's build succeeds, it is automatically merged and out-dates all other 8 PRs. All of them will be updated, one of them merged and it'll outdate 7 PRs, etc. For 10 PRs you'll get (worst-case) 10 merge commits.

I agree with you. Sorry, ignore this suggesstion.

**Comment by bobvanderlinden** [comment]
> What kind of concerns do you have for that?

This requires more requests upfront. The requests are currently only executed within the repository queue. There are edge cases where I'm not sure it gives the desired/expected result.

Say you're reviewing 3 PRs and there are implicit dependencies between the PRs. You approve one, thinking that it would be merges first. It needs to be updated, so it runs the ci. Meanwhile you approve another pr that depends on the first. That or doesn't need updating the branch and thus will be merges immediately.

This is fine, because even though they implicitly depend on each other, they can be merges either way. However, I think there is something to be said about keeping the order the user approved/labeled PRs and updating/merging in that order.

A good argument can also be that PRs should not depend on each other implicitly and that you'd need to use something like dep to explicitly define dependencies.

This makes it not trivial to know which method is better.

**Comment by dkhmelenko** [comment]
I've got your point. What do you think about enabling this option from configuration .yaml file? Something like `merge-ready-immediately: true`. By default this parameter is false.

I see this adding some complexity, but having a queue for scheduling tasks is already a bit complex solution :) Therefore it would be nice to have an option to merge green PRs immediately.

**Comment by ostrolucky** [comment]
We have slightly similar request, but I didn't test this tool enough to know if it doesn't already do what we need. So, can anybody explain how is auto-merge right now determining order of merges? What I mean with this is what if at point that CI turns from red to green there are 3 PR candidates to merge?

I would expect following config
```
requiredLabels:
- "release/urgent"
- "release/anytime"
```

to merge PRs marked as "release/urgent" first when there are 3 green PRs with these labels

**Comment by bobvanderlinden** [comment]
Currently they'll be handled in the order they came in. If PR A, B and C will go green in that order, they'll be handled in that same order. Merges will happen in that same order.

I'm not sure if an urgent label is that useful. What is the usecase where you'd want a PR to be merged before others and those PRs have gone green at the same time?

Note that the PRs will always go green in a specific order. When one nonurgent PR goes green, no other PR is green yet, so that PR will be merged anyway. It could only be applicable when updating PRs and waiting for their CI to pass while other more-urgent PRs go green.

**Comment by ostrolucky** [comment]
True. So I guess what I need here is to ignore green status of PRs if there is open PR marked as urgent

**Comment by bobvanderlinden** [comment]
So, if there is a single PR that is not green that has an urgent label, all other PRs won't be able to be merged? That sounds very confusing. Maybe I don't understand your usecase yet. How many PRs per hour are you merging? How long does it take for CI to go green?

**Comment by ostrolucky** [comment]
CI takes between 30-60minutes. When PR is merged, all other PRs need to be rebased on master and rebuilt. This means if we don't merge urgent PR first and there are 4 other PRs marked as released, it can take up to 5 hours to merge everything.
